enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,447 @@
+active/active on multiple nodes (masters).
+While this sounds complex, this comes due to abusing the deployment having different subnet IPs
+on the same segment and expect it to not go through the router, so we need to make symmetric traffic.
+
+This implies to create an IP table on each node where FRR will be running, that will enforce the describe behavior:
+```bash
+ip rule add from $VIP to $NETWORK_CIDR lookup <number value>
+ip route add default via $GATEWAY table <number value>
+```
+
+
+#### frr-monitor
+
+This container will have in charge:
+- Generate `frr.conf` configuration file so FRR can work correctly
+- Watch for configuration changes or events that require FRR daemons to stop or reload.
+
+The code will live in baremetal-runtimecfg among other monitors (HAproxy, Keepalived, etc).
+
+### Load balanced VIPs
+
+With Keepalived, the VIPs were located on one (elected) host (using [VRRP](https://fr.wikipedia.org/wiki/Virtual_Router_Redundancy_Protocol)) and HAproxy on that node
+would distribute the traffic across available control plane nodes.
+
+With BGP anycast, the VIPs will reside on all control plane hosts, and upstream routers (e.g. ToR) will define the best route (according to BGP protocol) and depending on how the routers
+are configured, will load-balance the traffic depending on how the infrastructure network is configured.
+For example, if the ToR router is a Linux machine, we can control which hash policy to use for multipath routes with
+`fib_multipath_hash_policy` in sysctl and choose if we want L3 or L4 load balancing. If a Cisco router is used, the `maximum-paths` option can be helpful. It really
+depends on the upstream router and BGP infrastructure that is in place, and not managed by OpenShift.
+
+
+### baremetal-runtimecfg
+
+[baremetal-runtimecfg](https://github.com/openshift/baremetal-runtimecfg) is used for rendering the manifests and templates of the machine-config-operator project, it also supports runtime update of the configuration templates (e.g: update FRR config in case a new BGP peer added to the Failure Domain) based on the current system status.
+
+The next capabilities are supported by the `baremetal-runtimecfg`:
+- `renders templates` using [values provided at command line](https://github.com/openshift/baremetal-runtimecfg/blob/master/cmd/runtimecfg/render.go), parameters retrieved from the cluster (e.g: control plane nodes) and
+  - api-vips
+  - ingress-vips
+- `frr monitor`: verify that the desired configuration is synced with the cluster. As seen before, it's used from a side-car container in the FRR pod.
+
+### API Extensions
+
+None.
+
+### Implementation Details/Notes/Constraints
+
+#### MetalLB
+
+MetalLB is a load-balancer implementation for bare metal Kubernetes clusters, using standard routing protocols and backed by FRR when using BGP as a protocol.
+
+For now, we recommend separating the MetalLB deployment from the workers where Ingress Controller runs.
```

Comment:
Can we not just use MetalLB for Ingress?

Reply:
MetalLB for Ingress is a great idea and I'm all for it. I think we should talk about it here (like I started a bit).
However, I believe this is a **separate** effort to make it happen from this RFE, as MetalLB can't be deployed on day 1.

I think that using MetalLB for Ingress can be done without this RFE, even today, and should be a documented and tested procedure at some point. But I feel like these are 2 different efforts, simply because this can happen today without this RFE.

Reply:
Why can't MetalLB be deployed on day 1?

You mention the problem which is giving me pause below: MetalLB *can't* run on any node where we're running this. That sounds like a really tricky upgrade problem which we could avoid by not doing it in the first place.

Reply:
> Why can't MetalLB be deployed on day 1?

It's not an operator that is installed during a cluster deploy, it has to be manually installed after a deployment and as you know it requires k8s APi to function.
I agree that MetalLB would be really nice candidate to run Ingress, but I don't think that's in the scope of this RFE, as we could already do that today without this RFE; isn't?

Reply:
Agreed, managing Ingress with MetalLB should be out of scope for this enhancement proposal.